Doc translate (#2)
shocking-rodents/magic-settings: README.md

# Magic-settings

## Installation

```bash

> $ pip install magic-settings

Reviewer: А можно не только к переводу придираться? А то маркдаунлинту [что-то тут не нравится](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md014)

```

Using settings from `yaml` file

```bash
$ pip install magic-settings[yaml]
```

## Initialization

> ### Project settings class declaration

Reviewer: ```suggestion
### Project settings class declaration

```

[MD022 - Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md022)

```python
from magic_settings import BaseSettings, Property

class MySettings(BaseSettings):
    VERSION = Property(types=str)
    PROJECT_DIR = Property(types=str)

> ```

Reviewer 2: Что означает «значение аттрибута класса `Property`»? Я в русской версии не очень понимаю, что это такое

> Class ```Property``` is a descriptor with following parameters:

Reviewer: MD031 - Fenced code blocks should be surrounded by blank lines

```suggestion

Class ```Property``` is a descriptor with following parameters:
```

> - ***types*** - Type list or type of ```value```. Raises ```ValueError``` if ```value``` is not one of the ```types```.

Reviewer 2: Он точно может быть списком? Если это передаётся в `isinstance`, то параметр обязан быть tuple.

Reviewer 2: type of `value` or a tuple of possible types. It is a `ValueError` if `value` is not one of the `types`.

> - ***validators*** - List of ```callable``` objects each of which is consistently applied to ```value```.  Raises ```ValueError``` if ```value``` does not pass at least one of the validations.

Reviewer 2: consistently -> successively

> - ***choices*** - List of any objects. If ```value``` not contained in ```choices``` - raises ```ValueError```. When using this parameter, parameters  ```types``` and ```validators``` are ignored.

Reviewer 2: if `value` is not in `choices`

- ***default*** - Sets the default value of the class attribute ```Property```.

> - ***converts*** - List of ```callable``` objects. It is a chain of transformations that are consistently applied to the ```value``` and overwrite it each time. It applies to ```value``` only if ```value``` is a string. Raises ```ValueError``` if ```value``` at least one of the transformations failed to apply.

Reviewer 2: successively

> 

Reviewer 2: Что такое объект внутри конструктора?

> ### Settings configuration

Reviewer: ```suggestion
### Settings configuration

```

> Settings configuration occur at the stage of creating a Settings object.

Reviewer 2: ```suggestion
Settings configuration occurs at the stage of creating a Settings object.
```

)
```

> ### Parameters

Reviewer: ```suggestion
### Parameters

```

> - ***modules***: List of python files to import with variables. Default - ```None```. 

Reviewer 2: ```suggestion
- ***modules***: List of Python modules with variables to import. Default ```None```. 
```

- ***prefix***: The prefix with which the environment variables are taken. Default - ```None```.

    _settings.py_

>     ```python

Reviewer: [MD031 - Fenced code blocks should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md031---fenced-code-blocks-should-be-surrounded-by-blank-lines)

```suggestion

    ```python
```

)
```

### Parameters

- ***modules***: List of Python modules with variables to import. Default ```None```.
- ***prefix***: The prefix with which the environment variables are taken. Default - ```None```.

    _settings.py_

    ```python
    class MySettings(BaseSettings):
        PSYDUCK = Property(types=str)
    ```

    _.env_

    ```dotenv
    MYPROJECT_PSYDUCK=Owowowow
    ```

    _some_other_place.py_

    ```python
    settings = MySettings(prefix='MYPROJECT')
    ```

    or

    ```python
    settings = MySettings(prefix='MYPROJECT_')
    ```

- ***dotenv_path***: Path to env-file. Default - ```None```. Using for exporting variables from env-file to environment. If ```dotenv_path``` is ```None``` -  walking up the directory tree looking for the specified file - called ```.env``` by default.

> - ***override_env***: ```True``` - override existing system environment variables, ```False``` - do not override. Default - ```False```.

Reviewer: не очень понятно, что оверрайдится - создается впечатлене, что энвы ОС

Author: load_dotenv - ```Reads the key,value pair from .env file and adds them to environment variable.```

Author: да, это энвы ос
у нас есть 2 стадии - загрузка энвов и выгрузка, так что все ок

Reviewer: тогда напиши, что это из .env файла

Author: написал

)
```

### Parameters

> - ***modules***: List of Python modules with variables to import. Default ```None```. 

Reviewer: MD009/no-trailing-spaces: Trailing spaces
```suggestion
- ***modules***: List of Python modules with variables to import. Default ```None```.
```

- ***prefix***: The prefix with which the environment variables are taken. Default - ```None```.

    _settings.py_

>     ```python

Reviewer: Здесь и ниже MD031/blanks-around-fences

```suggestion

    ```python
```

    settings = MySettings(prefix='MYPROJECT_')
    ```

>     

Reviewer: MD009/no-trailing-spaces

```suggestion

```

- ***dotenv_path***: Path to env-file. Default - ```None```. Using for exporting variables from env-file to environment. If ```dotenv_path``` is ```None``` -  walking up the directory tree looking for the specified file - called ```.env``` by default.
- ***override_env***: ```True``` - override existing system environment variables, ```False``` - do not override. Default - ```False```. 
- ***yaml_settings_path***: Path to yaml config file. Default - ```None```.
- ***use_env***: ```True``` - use environment variables. Default - ```True```.

> ### Exceptions

Reviewer: MD022/blanks-around-headings/blanks-around-headers

```suggestion
### Exceptions

```

***ValueError***: If ***modules*** type is not ```list``` or ```NoneType``` and if type of element in ***modules*** is not ```ModuleType```.

> Settings loading

Reviewer: [MD003 - Heading style](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md003---heading-style)
```suggestion
## Settings loading
```

> ----------------

Reviewer: [MD003 - Heading style](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md003---heading-style)
MD022 - Headings should be surrounded by blank lines

```suggestion

```

> Loading settings can be initiated anywhere in the project.

Reviewer: MD031/blanks-around-fences: Fenced code blocks should be surrounded by blank lines

```suggestion
Loading settings can be initiated anywhere in the project.

```

```python
from where_your_settings import settings

settings.init()
```

> If called again, it goes through the configuration files and update properties.

Reviewer: MD031/blanks-around-fences: Fenced code blocks should be surrounded by blank lines
```suggestion

If called again, it goes through the configuration files and update properties.
```

> Settings priority

Reviewer: [MD003 - Heading style](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md003---heading-style)
```suggestion
## Settings priority
```

> -----------------

Reviewer: ```suggestion

```

In case of intersection of settings the following priority will be applied:
_my_module_ -> _my_awesome_module_ -> _.env_ -> _settings.yaml_

> ```python

Reviewer: Слишком много такого ниже
[MD031 - Fenced code blocks should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md031---fenced-code-blocks-should-be-surrounded-by-blank-lines)

    settings = MySettings(prefix='MYPROJECT_')

>     ```

Reviewer: ```suggestion
    ```

```

>     

Reviewer: ```suggestion

```

> - ***dotenv_path***: Path to env-file. Default - ```None```. Using for exporting variables from env-file to environment. If ```dotenv_path``` is ```None``` -  walking up the directory tree looking for the specified file - called ```.env``` by default.

Reviewer: ```suggestion

- ***dotenv_path***: Path to env-file. Default - ```None```. Using for exporting variables from env-file to environment. If ```dotenv_path``` is ```None``` -  walking up the directory tree looking for the specified file - called ```.env``` by default.
```

PSYDUCK: yaml
```

> Examples:

Reviewer: [MD003 - Heading style](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md003---heading-style)
[MD026 - Trailing punctuation in heading](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md026)
```suggestion
## Examples

```

> --------

Reviewer: ```suggestion
```

# PSYDUCK = 'yaml'
```

> Temporary Property redefinition

Reviewer 2: override?

# PSYDUCK = 'yaml'
```

> Temporary Property override

Reviewer: ```suggestion
## Temporary Property override

```

> ---------------------------

Reviewer: ```suggestion
```

> _my_module.py_

Reviewer: [MD036 - Emphasis used instead of a heading](https://github.com/DavidAnson/markdownlint/blob/v0.13.0/doc/Rules.md#md036)

# PIKACHU = 'Psyduck_is_not_fine'
```

Method ```temp_set_attributes``` is not thread-safe.

> ## Settings list

Reviewer: ```suggestion
## Settings list

```
MD022/blanks-around-headings/blanks-around-headers: Headings should be surrounded by blank lines

You can use methods `to_dict()`, `to_json()` to get list of settings:
```python
from magic_settings import BaseSettings, Property

> class MySettings(BaseSettings):

Reviewer: Там ниже под объявлением класса MD009/no-trailing-spaces, которую я не могу исправить. А еще мне немного странным кажется вывод больших кусков информации в комментарии. Может, оформить это как вывод pprint?

Reviewer: Лишние пробелы не убрал, и про комментарии тоже не ответил

settings.init()

with settings.temp_set_attributes(PSYDUCK='I_am_ok', PIKACHU='Psyduck_is_ok'):
    print(settings.PSYDUCK) # 'I_am_ok'
    print(settings.PIKACHU) # 'Psyduck_is_ok'

>     

Reviewer: MD009/no-trailing-spaces: Trailing spaces

# PIKACHU = 'Psyduck_is_not_fine'
```

Method ```temp_set_attributes``` is not thread-safe.

## Settings list

> You can use methods `to_dict()`, `to_json()` to get list of settings:

Reviewer 2: ```suggestion
You can use methods `to_dict()`, `to_json()` to get current settings:
```

Всё-таки dict и json это не list

# }
```

> ## Validation

Reviewer: ```suggestion
## Validation

```

> It is recommended to use following `BaseSettings` class methods during redefinition `update_settings_from_source` method:

Reviewer: ```suggestion
It is recommended to use following `BaseSettings` class methods during redefinition `update_settings_from_source` method:

```

>     

Reviewer: ```suggestion

```

> 1. `pre_validate` - check for the presence of types in` Property`, check for compliance with the type `Property` in possible values ​​of` Property.choices`, check for default values ​​for type matching.

Reviewer: MD038/no-space-in-code: Spaces inside code span elements

```suggestion
1. `pre_validate` - check for the presence of types in `Property`, check for compliance with the type `Property` in possible values ​​of `Property.choices`, check for default values ​​for type matching.
```

> 1. `pre_validate` - check for the presence of types in` Property`, check for compliance with the type `Property` in possible values ​​of` Property.choices`, check for default values ​​for type matching.

Reviewer 3: ```suggestion
1. `pre_validate` - check for presence of types in` Property`, check for compliance with the type `Property` in possible values ​​of` Property.choices`, check for default values ​​for type matching.
```

2. `post_validate` - check if each `Property` is assigned a value.
  

> 

Reviewer: MD012/no-multiple-blanks: Multiple consecutive blank lines 
```suggestion
```

> Dynamic settings

Reviewer: ```suggestion
## Dynamic settings

```

> ----------------

Reviewer: ```suggestion
```

> ### Definition of class working with settings source:

Reviewer: MD022/blanks-around-headings/blanks-around-headers: Headings should be surrounded by blank lines 
MD026/no-trailing-punctuation: Trailing punctuation in heading

```suggestion
### Definition of class working with settings source

```

# }
```

## Validation

It is recommended to use following `BaseSettings` class methods during redefinition `update_settings_from_source` method:

> 1. `pre_validate` - check for presence of types in `Property`, check for compliance with the type `Property` in possible values ​​of `Property.choices`, check for default values ​​for type matching.

Reviewer 2: check that types are configured correctly; check that the values from `choices` and the `default` pass the type check.

> 1. `pre_validate` - check for the presence of types in `Property`, check for compliance with the type `Property` in possible values ​​of `Property.choices`, check for default values ​​for type matching.

Reviewer 3: ```suggestion
1. `pre_validate` - check for presence of types in `Property`, check for compliance with the type `Property` in possible values ​​of `Property.choices`, check for default values ​​for type matching.
```

2. `post_validate` - check if each `Property` is assigned a value.
  
## Dynamic settings

> ### Definition of class working with settings source

Reviewer 2: ```suggestion
### Implementing a custom dynamic settings source
```

        return super().update_config(**kwargs)
```

### Definition of project`s dynamic settings class

```python
class MyDynamicSettings(BaseDynamicSettingsDict):
    JIGGLYPUFF = Property(types=str)
```

> ### Initialization of dynamic settings class instance

Reviewer 2: ```suggestion
### Dynamic Settings Initialization
```

```python
loop = asyncio.get_event_loop()
dynamic_settings = MyDynamicSettings(loop=loop, update_period=5, task_retries_number=5)
```

- ***update_period***: settings`s update period from source in seconds.
- ***task_retries_number***: the number of attempts to update the settings when an exception occurred before stopping the task.

### Dynamic settings update

#### The only settings update 

```python
await dynamic_settings.update_settings_from_source()
```

> #### Running infinity settings update loop:

Reviewer 2: ```suggestion
#### Starting the update loop:
```

```python
await dynamic_settings.start_update()
```

> #### Stop infinity settings update loop:

Reviewer 2: ```suggestion
#### Stopping the update loop:
```

        return super().update_config(**kwargs)
```

### Definition of project`s dynamic settings class

```python
class MyDynamicSettings(BaseDynamicSettingsDict):
    JIGGLYPUFF = Property(types=str)
```

### Initialization of dynamic settings class instance

```python
loop = asyncio.get_event_loop()
dynamic_settings = MyDynamicSettings(loop=loop, update_period=5, task_retries_number=5)
```

> - ***update_period***: settings`s update period from source in seconds.

Reviewer 2: ```suggestion
- ***update_period***: settings update period from source in seconds.
```

- ***task_retries_number***: the number of attempts to update the settings when an exception occurred before stopping the task.

### Dynamic settings update

> #### The only settings update 

Reviewer 2: ```suggestion
#### Updating settings only once:
```

settings = MySettings(modules=[my_module, my_awesome_module], dotenv_path='/path/to/dotenv')
# PSYDUCK = 'env'
```
```python
settings = MySettings(modules=[my_module, my_awesome_module], dotenv_path='/path/to/dotenv', use_env=False)
# PSYDUCK = 'two'
```
```python
settings = MySettings(modules=[my_module, my_awesome_module], dotenv_path='/path/to/dotenv', yaml_settings_path='/path/to/yaml/settings.yaml')
# PSYDUCK = 'yaml'
```

Temporary Property redefinition
----------------------------------
_my_module.py_
```python
PIKACHU = 'Psyduck_is_not_fine'
PSYDUCK = 'Owowowow'

```
```python
from my_project import my_module
from my_config import MySettings

class MySettings(BaseSettings):
    PSYDUCK = Property(types=str)
    PIKACHU = Property(types=str)

settings = MySettings(modules=[my_module])
settings.init()

with settings.temp_set_attributes(PSYDUCK='I_am_ok', PIKACHU='Psyduck_is_ok'):
    print(settings.PSYDUCK)
    print(settings.PIKACHU)
    # PSYDUCK='I_am_ok'
    # PIKACHU='Psyduck_is_ok'

print(settings.PSYDUCK)
print(settings.PIKACHU)
# PSYDUCK = 'Owowowow'
# PIKACHU = 'Psyduck_is_not_fine'
```

Method ```temp_set_attributes``` is not thread-safe.

## Settings list

> You can use methods `to_dict()`, `to_json` to get list of settings:

Reviewer 4: to_json - тоже метод
